### PR TITLE
Exclude config and diagnostic entities from scenes

### PR DESCRIPTION
**Pull request by emontnemery**
## Proposed change

Exclude config and diagnostic entities from scenes when adding a device to the scene.

Config and diagnostic entities can still be added by adding individual entities to the scene.

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by emontnemery** [comment]
Maybe this can be handled here instead: https://github.com/home-assistant/frontend/blob/0bec9fa0af022cf4761ef42964107f95b62f303d/src/panels/config/scene/ha-scene-editor.ts#L531-L554

However, I'm not sure that code is correct; it doesn't seem to handle entities being removed or changed, only entities being added.

**Review by frenck — COMMENTED** [review]
How can I override this in case I still want to include it?

**Comment by emontnemery** [comment]
@frenck you can include a diagnostic or config entity manually, this just excludes config and diagnostic entities when adding a device.

**Comment by frenck** [comment]
Alright, sounds perfect, thanks 👍

File: src/panels/config/scene/ha-scene-editor.ts [hunk 1]
```diff
@@ -158,7 +158,16 @@ export class HaSceneEditor extends SubscribeMixin(
...
-          const deviceEntities: string[] = deviceEntityLookup[deviceId] || [];
+          const filteredEntityReg = this._entityRegistryEntries.filter(
+            (entityReg) =>
+              (deviceEntityLookup[device.id] || []).includes(
+                entityReg.entity_id
+              ) && entityReg.entity_category === null
+          );
+          const deviceEntities = filteredEntityReg.map(
+            (entityReg) => entityReg.entity_id
+          );
```
Comment (bramkragten) [inline src/panels/config/scene/ha-scene-editor.ts:169]:
```suggestion
          const deviceEntities: string[] = [];
          this._entityRegistryEntries.forEach(
            (entityReg) => {
              if((deviceEntityLookup[device.id] || []).includes(
                entityReg.entity_id
              ) && entityReg.entity_category === null) {
              	deviceEntities.push(entityReg.entity_id)
              }
            }
          );
```

**Comment by bramkragten** [comment]
> Maybe filtering out config and diagnostic entities can be handled here instead:
> 
> https://github.com/home-assistant/frontend/blob/0bec9fa0af022cf4761ef42964107f95b62f303d/src/panels/config/scene/ha-scene-editor.ts#L531-L554
> 
> However, I'm not sure that code is correct; it doesn't seem to handle entities being removed or changed, only entities being added.

Yes, I would make sure only entities without an entity category are added to `_deviceEntityLookup`, then the rest of the code can stay the same.

File: src/panels/config/scene/ha-scene-editor.ts [hunk 2]
```diff
@@ -664,6 +678,9 @@ export class HaSceneEditor extends SubscribeMixin(
     this._entities = this._entities.filter(
       (entityId) => entityId !== deleteEntityId
     );
+    this._single_entities = this._entities.filter(
```
Comment (bramkragten) [inline src/panels/config/scene/ha-scene-editor.ts:681]:
```suggestion
    this._single_entities = this. _single_entities.filter(
```

File: src/panels/config/scene/ha-scene-editor.ts (hunk 1 above)
Comment (emontnemery) [reply src/panels/config/scene/ha-scene-editor.ts]: No longer relevant

File: src/panels/config/scene/ha-scene-editor.ts [hunk 3]
```diff
@@ -520,23 +522,22 @@ export class HaSceneEditor extends SubscribeMixin(
     }
 
     if (changedProps.has("_entityRegistryEntries")) {
+      this._deviceEntityLookup = {};
```
Comment (bramkragten) [inline src/panels/config/scene/ha-scene-editor.ts:525]: remove `@state()` from `_deviceEntityLookup` so you don't trigger a new update

Comment (bramkragten) [reply src/panels/config/scene/ha-scene-editor.ts:525]: Or move this logic to the subscription code

File: src/panels/config/scene/ha-scene-editor.ts [hunk 4]
```diff
@@ -664,6 +678,9 @@ export class HaSceneEditor extends SubscribeMixin(
     this._entities = this._entities.filter(
       (entityId) => entityId !== deleteEntityId
     );
+    this._single_entities = this. _single_entities.filter(
```
Comment (bramkragten) [inline src/panels/config/scene/ha-scene-editor.ts:681]:
```suggestion
    this._single_entities = this._single_entities.filter(
```

File: src/panels/config/scene/ha-scene-editor.ts [hunk 5]
```diff
@@ -827,7 +844,7 @@ export class HaSceneEditor extends SubscribeMixin(
       }
 
       output[entityReg.entity_id] = {
-        entity_only: !this._devices.includes(entityReg.device_id!),
+        entity_only: this._single_entities.includes(entityReg.entity_id),
```
Comment (bramkragten) [inline src/panels/config/scene/ha-scene-editor.ts:847]: Instead of going over `this._entityRegistryEntries`, go over `this._single_entities`

**Review by zsarnett — APPROVED** [review]

**Pull request merged**